### PR TITLE
2.x: start reimplementing Maybe with up-to-date javadoc, tests and algo

### PR DESCRIPTION
**Pull request by akarnokd**
This PR reimplements the `Maybe` with more up-to-date algorithms (the original was a copy-paste of some older Single and Completable code it seems) and adds javadoc and tests along the way. Not all operators have been rewritten yet (and no placeholders). Early adopters should refer to the `toX` methods on the base reactive types. The PR also adds the `RxJavaPlugins` hooks to support `Maybe`.

In addition, I've fixed javadoc mistakes in the other base reactive types (referring to the wrong types).

### CODE REVIEW

File: src/main/java/io/reactivex/Maybe.java [hunk 1]
```diff
@@ -13,838 +13,797 @@
...
+     */
+    @SchedulerSupport(SchedulerSupport.NONE)
+    public static <T> Maybe<T> fromAction(final Action run) {
+        ObjectHelper.requireNonNull(run, "run is null");
+        return RxJavaPlugins.onAssembly(new MaybeFromAction<T>(run));
+    }
+    
+    /**
+     * Returns a {@link Maybe} that invokes passed function and emits its result for each new MaybeObserver that subscribes
+     * while considering {@code null} value from the callable as indication for valueless completion.
```
Comment (JakeWharton) [inline src/main/java/io/reactivex/Maybe.java:184]: nice

File: src/main/java/io/reactivex/MaybeObserver.java [hunk 2]
```diff
@@ -14,13 +14,54 @@
...
+     * Provides the MaybeObserver with the means of cancelling (disposing) the
+     * connection (channel) with the Maybe in both
+     * synchronous (from within {@code onSubscribe(Disposable)} itself) and asynchronous manner.
+     * @param d the Disposable instance whose {@link Disposable#dispose()} can
+     * be called anytime to cancel the connection
+     */
     void onSubscribe(Disposable d);
 
+    /**
+     * Notifies the SingleObserver with a single item and that the {@link Single} has finished sending
```
Comment (JakeWharton) [inline src/main/java/io/reactivex/MaybeObserver.java]: nit: MaybeObserver

File: src/main/java/io/reactivex/MaybeObserver.java [hunk 3]
```diff
@@ -14,13 +14,54 @@
...
+     * push-based notifications.
+     * <p>
+     * The {@link Single} will not call this method if it calls {@link #onError}.
```
Comment (JakeWharton) [inline src/main/java/io/reactivex/MaybeObserver.java]: nit: Maybe

File: src/main/java/io/reactivex/MaybeObserver.java (hunk 2 above)
Comment (JakeWharton) [inline src/main/java/io/reactivex/MaybeObserver.java]: also Maybe not Single

File: src/main/java/io/reactivex/MaybeObserver.java [hunk 4]
```diff
@@ -14,13 +14,54 @@
...
+     * 
+     * @param value
+     *          the item emitted by the Single
+     */
     void onSuccess(T value);
 
-    void onComplete();
-
+    /**
+     * Notifies the SingleObserver that the {@link Single} has experienced an error condition.
```
Comment (JakeWharton) [inline src/main/java/io/reactivex/MaybeObserver.java]: MaybeObserver & Maybe here and next line down

File: src/main/java/io/reactivex/internal/fuseable/HasUpstreamObservableSource.java [hunk 5]
```diff
@@ -11,15 +11,21 @@
...
+ * via {@link #source()} method.
+ *
+ * @param <T> the value type
+ */
+public interface HasUpstreamObservableSource<T> {
     /**
      * Returns the upstream source of this Observable.
      * <p>Allows discovering the chain of observables.
-     * @return the source ObservableConsumable
+     * @return the source SingleSource
```
Comment (JakeWharton) [inline src/main/java/io/reactivex/internal/fuseable/HasUpstreamObservableSource.java]: ObservableSoure

Comment (akarnokd) [reply src/main/java/io/reactivex/internal/fuseable/HasUpstreamObservableSource.java]: Fixing...

File: src/main/java/io/reactivex/internal/operators/maybe/MaybeFromObservable.java [hunk 6]
```diff
@@ -10,47 +10,107 @@
...
-            public void onNext(T value) {
-                last = value;
+        }
+
+        @Override
+        public void onNext(T t) {
+            if (!isDisposed()) {
+                if (value != null) {
+                    get().dispose();
+                    onError(new IndexOutOfBoundsException("The source Publisher produced more than one item"));
```
Comment (JakeWharton) [inline src/main/java/io/reactivex/internal/operators/maybe/MaybeFromObservable.java]: Publisher -> Observable

File: src/main/java/io/reactivex/internal/operators/maybe/MaybeFromObservable.java [hunk 7]
```diff
@@ -10,47 +10,107 @@
...
-        upstream.subscribe(new Observer<T>() {
-            T last;
-            @Override
-            public void onSubscribe(Disposable d) {
-                s.onSubscribe(d);
+    protected void subscribeActual(MaybeObserver<? super T> observer) {
+        source.subscribe(new FromPublisherToMaybeObserver<T>(observer));
+    }
+    
+    static final class FromPublisherToMaybeObserver<T>
```
Comment (JakeWharton) [inline src/main/java/io/reactivex/internal/operators/maybe/MaybeFromObservable.java]: not that it really matters, but FromPublisherToMaybeObserver -> FromObservableToMaybeObserver

File: src/main/java/io/reactivex/internal/operators/maybe/MaybeToCompletable.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+package io.reactivex.internal.operators.maybe;
+
+import io.reactivex.*;
+import io.reactivex.disposables.Disposable;
+import io.reactivex.internal.disposables.DisposableHelper;
+import io.reactivex.internal.fuseable.HasUpstreamMaybeSource;
+
+/**
+ * Wraps a MaybeSource and exposes its onSuccess and onError signals and signals
+ * NoSuchElementException for onComplete.
```
Comment (JakeWharton) [inline src/main/java/io/reactivex/internal/operators/maybe/MaybeToCompletable.java:23]: This is a lot more clear as to the behavior than `toCompletable()`'s javadoc.

**Comment by JakeWharton** [comment]
:+1:

**Comment by akarnokd** [comment]
Thanks @JakeWharton !

**Pull request merged**